addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java

```diff
@@ -0,0 +1,297 @@
+
+    /**
+     * Additional meter listeners to get received meter values.
+     */
+    private final List<P1TelegramListener> meterListeners = new ArrayList<>();
+
+    /**
+     * Serial Port Manager.
+     */
+    private final SerialPortManager serialPortManager = new SerialPortManager() {
+    };
+
+    /**
+     * The dsmrDevice managing the connection and handling telegrams.
+     */
+    private @Nullable DSMRDevice dsmrDevice;
+
+    /**
+     * Long running process that controls the DSMR device connection.
+     */
+    private @Nullable DSMRDeviceRunnable dsmrDeviceRunnable;
+    /**
+     * Thread for {@link DSMRDeviceRunnable}. A thread is used because the {@link DSMRDeviceRunnable} is a blocking
+     * process that runs as long as the thing is not disposed.
+     */
+    private @Nullable Thread dsmrDeviceThread;
+    /**
+     * Watchdog to check if messages received and restart if necessary.
+     */
+    private @Nullable ScheduledFuture<?> watchdog;
+    /**
+     * Number of nanoseconds after which a timeout is triggered when no messages received.
+     */
+    private long receivedTimeoutNanos;
+
+    /**
+     * Timestamp in nanoseconds of last P1 telegram received
+     */
+    private long telegramReceivedTimeNanos;
+
+    /**
+     * Constructor
+     *
+     * @param bridge the Bridge ThingType
+     */
+    public DSMRBridgeHandler(Bridge bridge) {
+        super(bridge);
+    }
+
+    /**
+     * The {@link DSMRBridgeHandler} does not support handling commands.
+     *
+     * @param channelUID the {@link ChannelUID} of the channel to which the command was sent
+     * @param command the {@link Command}
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // DSMRBridgeHandler does not support commands
+    }
+
+    /**
+     * Initializes this {@link DSMRBridgeHandler}.
+     *
+     * This method will get the corresponding configuration and initialize and start the corresponding
+     * {@link DSMRDevice}.
+     */
+    @Override
+    public void initialize() {
+        DSMRDeviceConfiguration deviceConfig = getConfigAs(DSMRDeviceConfiguration.class);
+
+        logger.trace("Using configuration {}", deviceConfig);
+        updateStatus(ThingStatus.UNKNOWN);
+        receivedTimeoutNanos = TimeUnit.SECONDS.toNanos(deviceConfig.receivedTimeout);
+        try {
+            DSMRDevice dsmrDevice = createDevice(deviceConfig);
+            // Start time monitoring.
+            resetLastReceivedState();
+            this.dsmrDevice = dsmrDevice; // otherwise Eclipse will give a null pointer error on the next line :-(
+            dsmrDeviceRunnable = new DSMRDeviceRunnable(dsmrDevice, this);
+            dsmrDeviceThread = new Thread(dsmrDeviceRunnable);
+            dsmrDeviceThread.start();
+            watchdog = scheduler.scheduleWithFixedDelay(this::alive, receivedTimeoutNanos, receivedTimeoutNanos,
```

Comment:
With such a short timing it might make sense to create your own scheduler and start both the `dsmrDeviceRunnable` and watch-dog in that, wdyt?

Because continuously (nano seconds scale sounds often) claiming a thread from the pool seems not good for the system as a whole.

Reply:
The timing is in seconds. But the time is used to compare against `System.nanoTime()`. I'll used nanoseconds so I don't have to convert the system nano time to seconds each time the comparison is done. This comparison is used to determine the action in case of problems. It's a scaled recovery process.